File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String, String> idName = new HashMap<>();
        List<String[]> answer = new ArrayList<>();
        String[] mention = new String[]{"님이 들어왔습니다.", "님이 나갔습니다."};
        
        for (String r : record) {
            String[] s = r.split(" ");
            if (s[0].equals("Leave")) {
                answer.add(new String[]{s[1], mention[1]});
            }
            else {
                idName.put(s[1], s[2]);
                if (s[0].equals("Enter")) {
                    answer.add(new String[]{s[1], mention[0]});
                }
            }
        }
        
        
        return answer.stream()
            .map(s -> idName.get(s[0]) + s[1])
            .toArray(String[]::new);
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):

    answer = []
    name = {}
    
    for r in record:
        command, *uid = r.split()

        if command == "Change":
            name[uid[0]] = uid[1]
        
        if command == "Enter":
            name[uid[0]] = uid[1]
            answer.append((uid[0], "님이 들어왔습니다."))
            
        elif command == "Leave":
            answer.append((uid[0], "님이 나갔습니다."))
    

    return list(map(lambda x: name[x[0]] + x[1], answer))
        
         